spark/src/main/java/org/apache/iceberg/spark/data/vectorized/IcebergArrowColumnVector.java

```diff
@@ -0,0 +1,753 @@
+    }
+
+    @Override
+    final double getDouble(int rowId) {
+      return dictionary.decodeToDouble(vector.get(rowId));
+    }
+  }
+
+  private class DecimalAccessor extends ArrowVectorAccessor {
+
+    private final IcebergArrowVectors.DecimalArrowVector vector;
+
+    DecimalAccessor(IcebergArrowVectors.DecimalArrowVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final Decimal getDecimal(int rowId, int precision, int scale) {
+      if (isNullAt(rowId)) {
+        return null;
+      }
+      return Decimal.apply(vector.getObject(rowId), precision, scale);
+    }
+  }
+
+  private class StringAccessor extends ArrowVectorAccessor {
+
+    private final IcebergArrowVectors.VarcharArrowVector vector;
+    private final NullableVarCharHolder stringResult = new NullableVarCharHolder();
+
+    StringAccessor(IcebergArrowVectors.VarcharArrowVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final UTF8String getUTF8String(int rowId) {
+      vector.get(rowId, stringResult);
+      if (stringResult.isSet == 0) {
+        return null;
+      } else {
+        return UTF8String.fromAddress(
+            null,
+            stringResult.buffer.memoryAddress() + stringResult.start,
+            stringResult.end - stringResult.start);
+      }
+    }
+  }
+
+  private class DictionaryStringAccessor extends ArrowVectorAccessor {
+
+    private final IntVector vector;
+
+    DictionaryStringAccessor(IntVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final UTF8String getUTF8String(int rowId) {
+      if (isNullAt(rowId)) {
+        return null;
+      }
+      Binary binary = dictionary.decodeToBinary(vector.get(rowId));
+      return UTF8String.fromBytes(binary.getBytesUnsafe());
+    }
+  }
+
+  private class FixedSizeBinaryAccessor extends ArrowVectorAccessor {
+
+    private final FixedSizeBinaryVector vector;
+
+    FixedSizeBinaryAccessor(FixedSizeBinaryVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final byte[] getBinary(int rowId) {
+      return vector.getObject(rowId);
+    }
+  }
+
+  private class BinaryAccessor extends ArrowVectorAccessor {
+
+    private final VarBinaryVector vector;
+
+    BinaryAccessor(VarBinaryVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final byte[] getBinary(int rowId) {
+      return vector.getObject(rowId);
+    }
+  }
+
+  private class DictionaryBinaryAccessor extends ArrowVectorAccessor {
+
+    private final IntVector vector;
+
+    DictionaryBinaryAccessor(IntVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final byte[] getBinary(int rowId) {
+      Binary binary = dictionary.decodeToBinary(vector.get(rowId));
+      return binary.getBytesUnsafe();
+    }
+  }
+
+  private class DateAccessor extends ArrowVectorAccessor {
```

Comment:
Have we tested using `IntVector` instead of `DateDayVector`?

It would simplify the decoding logic to always use an `IntVector`, and I'm not sure what the benefit of using a `DateDayVector` is. Isn't the underlying buffer identical?

Same logic applies for `TimeStampMicroTZVector` and `LongVector`.

Reply:
Filed https://github.com/apache/incubator-iceberg/issues/834 to follow up on this.